src/dkg.rs

```diff
@@ -0,0 +1,413 @@
+}
+
+impl<G: Group> DkgSharesCollector<G> {
+    /// Returns a secret share for a participant with the specified `participant_index`.
+    pub fn secret_share_for_participant(&self, participant_index: usize) -> SecretKey<G> {
+        self.dealer.secret_share_for_participant(participant_index)
+    }
+
+    /// Returns public participant information: participant's public polynomial,
+    /// proof of possession for the corresponding secret polynomial and decommitment.
+    pub fn public_info(&self) -> (Vec<G::Element>, ProofOfPossession<G>, [u8; 32]) {
+        let (public_polynomial, proof_of_possession) = self.dealer.public_info();
+        (
+            public_polynomial,
+            proof_of_possession.clone(),
+            self.decommitment,
+        )
+    }
+
+    /// Returns indices of parties whose shares were not provided.
+    pub fn missing_shares(&self) -> Vec<usize> {
+        self.shares_received
+            .iter()
+            .enumerate()
+            .filter_map(|(i, x)| if *x { None } else { Some(i) })
+            .collect()
+    }
+
+    /// Inserts secret share from participant with index `participant_index` and
+    /// checks if the share is valid.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if provided secret share doesn't lie on given public
+    /// polynomial or the polynomial doesn't correspond to the previous commitment.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.index` is greater or equal to number of participants,
+    /// which should never happen, unless DkgSharesCollector got corrupted.
+    pub fn insert_secret_share(
+        &mut self,
+        participant_index: usize,
+        secret_share: SecretKey<G>,
+        public_polynomial: &[G::Element],
+        proof_of_possession: &ProofOfPossession<G>,
+        decommitment: &[u8; 32],
+    ) -> Result<(), Error> {
+        if self.shares_received[participant_index] {
+            return Err(Error::DuplicitShare);
+        }
+
+        // verifies proof of possession of secret polynomial
+        let keyset =
+            match PublicKeySet::new(self.params, public_polynomial.to_vec(), proof_of_possession) {
+                Ok(keyset) => keyset,
+                Err(error) => return Err(Error::MalformedParticipantProof(error)),
+            };
+
+        let public_share = keyset.participant_key(self.index).unwrap();
+        if public_share.as_element() != G::mul_generator(secret_share.expose_scalar()) {
+            // point corresponding to the received secret share doesn't lie
+            // on the public polynomial
+            return Err(Error::InvalidSecret);
+        }
+
+        let mut hasher = Sha256::new();
+        let mut bytes = vec![0_u8; G::ELEMENT_SIZE];
+        G::serialize_element(&public_polynomial[0], &mut bytes);
+        hasher.update(&bytes);
+        hasher.update(decommitment);
+        let c: [u8; 32] = hasher.finalize().into();
+
+        if self.commitments[participant_index] != c {
```

Comment:
Nit: since this check does not involve any elliptic curve group operations (i.e., it's relatively inexpensive), it should be placed before the checks that do evolve EC operations, such as `PublicKeySet::new()`.